cmd/clusterctl/client/repository/repository_github.go

```diff
@@ -420,3 +416,12 @@ func (g *gitHubRepository) handleGithubErr(err error, message string, args ...in
 	}
 	return errors.Wrapf(err, message, args...)
 }
+
+func (g *gitHubRepository) checkGithubErr(err error, message string, args ...interface{}) (bool, error) {
+	retryError := g.handleGithubErr(err, message, args...)
```

Comment:
nit: because this is not generalized we might want to rename the `retryError` return var name to just `err`

Reply:
I see we're using `err` for the func input error object, so ignore that for the time being.

However, another point: this new `checkGithubErr` on top of `handleGithubErr` doesn't fully make sense to me, as we're testing if the `err` type is `github.RateLimitError` twice now. Is that intended?

Reply:
I think the entire usage surface area of `handleGithubErr` can be refactored, how about we revert all of these changes in this PR so we can simply focus on the improvements to pagination, and then one of us can do that refactor work in a follow-up PR? (happy to do it myself)

Reply:
basically it looks like `handleGithubErr` returns two types of errors, one for the rate limit and one for any other reason which might have caused the error.
```go
func (g *gitHubRepository) handleGithubErr(err error, message string, args ...interface{}) error {
	if _, ok := err.(*github.RateLimitError); ok {
		return errors.New("rate limit for github api has been reached. Please wait one hour or get a personal API token and assign it to the GITHUB_TOKEN environment variable")
	}
	return errors.Wrapf(err, message, args...)
}
```
But this check
```
retryError := g.handleGithubErr(err, message, args...)
// return immediately if we are rate limited
if _, ok := listReleasesErr.(*github.RateLimitError); ok {
	return false, retryError
}
return false, nil
```
 indicates that we are only surfacing the first error to the PollImmediate method, because that indicates a rate limit and we should stop the retry.
```
PollImmediate tries a condition func until it returns true, an error, or the timeout is reached.
```

Reply:
I guess we can return nil from `handleGithubErr` in the second case, and then we needn't check again for `github.RateLimitError` twice since it will only ever return error on rate limit

Reply:
It seems like an antipattern to mask an error as nil, but that's what seem to be doing [here](https://github.com/kubernetes-sigs/cluster-api/blob/main/cmd/clusterctl/client/repository/repository_github.go#L294)

Reply:
Sure, I can revert the changes